### PR TITLE
#80 - Initial support for auth-providers

### PR DESCRIPTION
**Pull request by javierarrieta**
This is an initial approach to work on auth-providers, still need to refresh tokens

### CODE REVIEW

File: client/src/main/scala/skuber/api/Configuration.scala [hunk 1]
```diff
@@ -143,23 +143,59 @@ object Configuration {
...
+              case _ => None
             }
-            case false => None
-          },
-          token = optionalValueAt(userConfig, "token"),
-          userName = optionalValueAt(userConfig, "username"),
-          password = optionalValueAt(userConfig, "password"))
+          }
+
+          val maybeAtuh = optionalValueAt[YamlMap](userConfig, "auth-provider") match {
```
Comment (doriordan) [inline client/src/main/scala/skuber/api/Configuration.scala]: I assume  'maybeAtuh' is a typo, should be 'maybeAuth'?

File: client/src/main/scala/skuber/api/Configuration.scala [hunk 2]
```diff
@@ -143,23 +143,59 @@ object Configuration {
...
+                case false => None
+              }
 
-        val k8sAuthInfoMap = topLevelYamlToK8SConfigMap("user", toK8SAuthInfo _)
+              val token = optionalValueAt(userConfig, "token")
+
+              val userName = optionalValueAt(userConfig, "username")
+              val password = optionalValueAt(userConfig, "password")
+
+              (
```
Comment (doriordan) [inline client/src/main/scala/skuber/api/Configuration.scala]: I think a pattern match approach would be clearer, something like
```
  (userName,password,token,jwt,clientCertificate,clientKey) match {
    case (Some(u), Some(p),_,_,_,_)  => BasicAuth(u,p)
    ...
  }
```

File: client/src/main/scala/skuber/api/package.scala [hunk 3]
```diff
@@ -51,11 +49,66 @@ package object client {
 
    case class Context(
      cluster: Cluster = Cluster(),
-     authInfo: AuthInfo = AuthInfo(),
+     authInfo: AuthInfo = NoAuth,
      namespace: Namespace = Namespace.default
    )
 
-   case class AuthInfo(
+   sealed trait AuthInfo
```
Comment (doriordan) [inline client/src/main/scala/skuber/api/package.scala]: I like this refactoring, the single AuthInfo class was getting cumbersome

File: client/src/main/scala/skuber/api/package.scala [hunk 4]
```diff
@@ -51,11 +49,66 @@ package object client {
...
+   }
+
+   final case class GcpAuth(accessToken: String) extends AuthProviderAuth {
+     override val name = "gcp"
+
+     override def toString =
+       """GcpAuth(accessToken=<redacted>)""".stripMargin
+   }
+
+   case class AuthInfo1(
```
Comment (doriordan) [inline client/src/main/scala/skuber/api/package.scala]: I assume `AuthInfo1` should be removed?

File: client/src/main/scala/skuber/api/Configuration.scala (hunk 1 above)
Comment (javierarrieta) [reply client/src/main/scala/skuber/api/Configuration.scala]: Yeah, thanks. Damn phat phingers

File: client/src/main/scala/skuber/api/package.scala (hunk 4 above)
Comment (javierarrieta) [reply client/src/main/scala/skuber/api/package.scala]: Yeah, I kept that to have easy the old code, should have removed before committing

**Comment by javierarrieta** [comment]
Hi @doriordan , changed the requested code, thanks for the review. I am unsure if it is worth merging now or wait until I implement the token refresh as it might change the API a bit? Not sure when you want to release

**Comment by doriordan** [comment]
Hi @javierarrieta if you are planning to add token refresh in the near future then I think it is best to hold off on the merge

**Comment by javierarrieta** [comment]
Cool, will do that now when I have a chance, wanted to be sure you were happy with the approach first

Thanks!

**Comment by javierarrieta** [comment]
Added refreshing of the Gcp token, not sure how oicd is refreshed. Refresh is done on demand with a hardcoded threshold of 30 seconds.

For a ouaht2 lib I implemented a state machine that will refresh the tokens, but might be too complex for this use case and also would need to change the api to `Future[String]`

Opinions @doriordan ?

File: client/src/main/scala/skuber/api/package.scala [hunk 5]
```diff
@@ -51,37 +52,115 @@ package object client {
...
+   }
+
+   final private[client] case class GcpRefresh(accessToken: String, expiry: Instant) {
+     def expired: Boolean = Instant.now.isAfter(expiry.minusSeconds(20))
+   }
+
+  private[client] object GcpRefresh {
+    implicit val gcpRefreshReads: Reads[GcpRefresh] = (
+      (JsPath \ "credential" \ "access_token").read[String] and
+      (JsPath \ "credential" \ "token_expiry").read[Instant]
```
Comment (javierarrieta) [inline client/src/main/scala/skuber/api/package.scala]: This should not be hardcoded, but don't know how to use JsPath to apply what there is in the config, I should evaluate the JsPath in these

```
        expiry-key: '{.credential.token_expiry}'
        token-key: '{.credential.access_token}'
```

**Comment by javierarrieta** [comment]
Hi @doriordan any stopper for this? I could remove the hardcode and oidc refresh later if you are happy with the approach

**Comment by doriordan** [comment]
Hi @javierarrieta I just want to have another look before I merge but I don't think there is any stopper for this

**Comment by javierarrieta** [comment]
Cool, thanks.

**Comment by nemo83** [comment]
@javierarrieta @doriordan can this be used to authenticate with EKS?

**Comment by javierarrieta** [comment]
Hi @nemo83 , haven't checked, hasn't been following this too much, any findings? Any way I can help?

**Comment by nemo83** [comment]
Compared to the java client and it doesn't seem that EKS is supported in this lib. Because of the very little code I had to write I decided to go ahead with the java lib instead. Thanks for replying to this.

**Comment by javierarrieta** [comment]
Sure, no worries, sorry for the very late reply. Any link to the java lib so maybe worth contributing to this repo?

**Comment by nemo83** [comment]
This is the k8s java client thread for the EKS auth impl:

https://github.com/kubernetes-client/java/issues/315

**Comment by doriordan** [comment]
Hi @javierarrieta sorry for late reply - thanks for the pointer and I'll take a look into what would be required to support EKS

**Comment by javierarrieta** [comment]
Is this related to https://github.com/doriordan/skuber/issues/173 ?

**Comment by doriordan** [comment]
@javierarrieta yes, need to revisit that and see if it is best to take that forward or a different take on it

**Pull request merged**